### PR TITLE
[16.0] google_gmail: avoid SMTPServerDisconnected errors

### PR DESCRIPTION
**Pull request by petrus-v**
## Description of the issue/feature this PR addresses:

This PR is about fixing this issue: https://github.com/odoo/odoo/issues/182777 to avoid SMTPServerDisconnected

## Current behavior before PR:

With some race condition got SMTPServerDisconnected

## Desired behavior after PR is merged:
Do not get SMTPServerDisconnected while quitting non connect session and avoid to get non connected session using google oauth2 connection.

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/google_gmail/models/google_gmail_mixin.py [hunk 1]
```diff
@@ -143,10 +143,14 @@ def _generate_oauth2_string(self, user, refresh_token):
         :return: The SASL argument for the OAuth2 mechanism.
         """
         self.ensure_one()
+        # token life is around 1 hour, add 5s threshold to gives
+        # time to contact google servers and avoid clock synchronization
+        # problems
+        threshold_seconds = 5
```
Comment (reth-odoo) [inline addons/google_gmail/models/google_gmail_mixin.py:149]: Can you specify that 5 seconds is the timeout for the request? Also would give it maybe a couple seconds of leeway to be safe.

Ideally would probably make it some named constant so we don't have to think to update it here.
e.g. `GMAIL_TOKEN_REQUEST_TIMEOUT = 5` at the top of the file

File: addons/google_gmail/tests/test_google_gmail.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+                "google_gmail_access_token": "fake_access_token",
+                "google_gmail_access_token_expiration": now_timestamp + 2,
+            }
+        )
+        with mock.patch(
+            "odoo.addons.google_gmail.models.google_gmail_mixin.GoogleGmailMixin._fetch_gmail_access_token",
+            return_value=("new-access-token", now_timestamp + 60 * 60)
+        ):
+            oauth2_string = mail_server._generate_oauth2_string("user-account", "refresh-token")
+        self.assertIn("new-access-token", oauth2_string)
```
Comment (reth-odoo) [inline addons/google_gmail/tests/test_google_gmail.py:32]: I think freezegun should work for this, would avoid relying on test timing.

File: addons/mail/tests/test_mail_mail.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
+from odoo.tests import TransactionCase
+from unittest import mock
+import smtplib
+
+
+class MailCase(TransactionCase):
+
+    def test_mail_send_missing_not_connected(self):
+        """This assume while calling self.env['ir.mail_server'].connect() it return
+        an disconnected smtp_session which is a non falsy object value"""
```
Comment (reth-odoo) [inline addons/mail/tests/test_mail_mail.py:10]: docstring should specify what we're testing first.
But I don't really get what your assumption is here, if you could maybe reword :)

```suggestion
        """Check getting a disconnected error while disconnecting doesn't raise.
        
        This assumes calling self.env['ir.mail_server'].connect() returns a disconnected smtp_session, which is a non-falsy object value."""
```

**Review by reth-odoo — COMMENTED** [review]
Hi there :wave:
Sounds good, thanks for the plentiful tests :D

You can target down to 16.0, seems stable-friendly enough

File: addons/google_gmail/models/google_gmail_mixin.py [hunk 4]
```diff
@@ -11,6 +11,18 @@
...
+The generated token is used until it's expiration (valid arround 1 hour)
+to open successive SMTP session
+"""
+GMAIL_TOKEN_VALIDITY_THRESHOLD = GMAIL_TOKEN_REQUEST_TIMEOUT + 1
+"""A threshold in seconds that we want to force the token
+to be re-generated because we are closed to its expiration time
+it gives a couple seconds of leeway to be safe to ensure we have
+time to open the new smtp session with the existing token.
+"""
+
```
Comment (reth-odoo) [inline addons/google_gmail/models/google_gmail_mixin.py:25]: No need to explain everything here, just the reason it's defined that way (unless it's arbitrary, like the first one)
```suggestion
GMAIL_TOKEN_REQUEST_TIMEOUT = 5

# seconds removed from end-of-validity datetime to take into account the time
# needed to renew the token
GMAIL_TOKEN_VALIDITY_THRESHOLD = GMAIL_TOKEN_REQUEST_TIMEOUT + 5

```

**Review by reth-odoo — COMMENTED** [review]
ok, just a nitpick for the tests otherwise lgtm :)

Thanks for the quick followup

EDIT: Also the smtp test seems to fail on runbot :x

**Comment by petrus-v** [comment]
@reth-odoo sorry for noise... I was lazy and give a try without local environment on v16.0... but at the end I setup it to test locally...

Thanks for all your reviews, now I'm happy with the diff and fixed test, I don't think the remaining tours error is related to this PR, I'd like to get some suggestion if so?

**Comment by petrus-v** [comment]
@reth-odoo , @tde-banana-odoo Should I do something with the runbot error, I've the feeling it's unrelated to this PR.

**Comment by reth-odoo** [comment]
@petrus-v indeed, our CI can be a bit capricious

I relaunched the test again, it's fine now :+1:

**Comment by petrus-v** [comment]
> @petrus-v indeed, our CI can be a bit capricious
> 
> I relaunched the test again, it's fine now 👍

Great, thanks !

**Comment by petrus-v** [comment]
@tde-banana-odoo, I know I'm a bit insistent, as far I understand the state of this PR it needs your review as mentioned by  @reth-odoo in above history

> @tde-banana-odoo not much more to say if you want to give this one a look

**Comment by reth-odoo** [comment]
@tde-banana-odoo :bubbles: up

**Comment by petrus-v** [comment]
kindly up!

**Comment by reth-odoo** [comment]
All this push notif stuff reminded me of this one :D
Sorry I was on vacation when you up-ed :')

@tde-banana-odoo :up:

File: addons/google_gmail/tests/test_google_gmail.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
+from odoo.tests.common import TransactionCase
+from unittest import mock
+from datetime import datetime
+from freezegun import freeze_time
+
+
+class TestIrMailServer(TransactionCase):
+
+    @freeze_time("2021-12-15 10:59:50")
+    def test_generate_oauth2_string_token_valid(self):
```
Comment (tde-banana-odoo) [inline addons/google_gmail/tests/test_google_gmail.py:10]: As most setup is common, can't you extract it into a setup and keep the test itself in test method(s) ?

File: addons/mail/tests/test_mail_mail.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+        Or here testing non recipients emails with non connected
+        smtp session, we won't get SMTPServerDisconnected that would
+        hide the other error that is raised earlier.
+        """
+        not_connected = smtplib.SMTP(local_hostname="fake-hostname.com", port=9999, timeout=1)
+        mail = self.env["mail.mail"].create({})
+        with mock.patch("odoo.addons.base.models.ir_mail_server.IrMailServer.connect", return_value=not_connected):
+            mail.send()
+        # if we get here SMTPServerDisconnected was not raised
+        self.assertEqual(mail.state, "exception")
```
Comment (tde-banana-odoo) [inline addons/mail/tests/test_mail_mail.py:24]: Well that's a bit of a weird test. Not sure what we are testing, can't we use mock tools and force a SMTPServerDisconnected error, and check it does not raise ?

**Review by tde-banana-odoo — COMMENTED** [review]
Thanks for the PR, added few comments, mainly on tests themselves :) 

cc @reth-odoo not sure if we retake it for the "test" part to close it ?

File: addons/google_gmail/tests/test_google_gmail.py (hunk 5 above)
Comment (petrus-v) [reply addons/google_gmail/tests/test_google_gmail.py]: done !

File: addons/mail/tests/test_mail_mail.py (hunk 6 above)
Comment (petrus-v) [reply addons/mail/tests/test_mail_mail.py:30]: We are now using a magick mock to raise an SMTPServerDisconnected on quit call and assert the quit method is called, much cleaner !

Also mock and assert the ignoring log message !

**Comment by petrus-v** [comment]
@tde-banana-odoo unit tests has been re-written using mock, does that suits to you?

**Comment by tde-banana-odoo** [comment]
It is freeze time, will try to not forget it. Can't we have reminders ?

**Comment by petrus-v** [comment]
> It is freeze time, will try to not forget it. Can't we have reminders ?

Not sure what should I do with that ! Do you want me to kindly ping you every days/weeks or move the PR against v17.0?

Actually, from my personal interest I would love to see it on v17.0.

File: addons/google_gmail/tests/test_google_gmail.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    def test_generate_oauth2_string_token_expired(self):
+        """Check that existing token is already expired
+        a new token is generated.
+        """
+        with mock.patch(
+            "odoo.addons.google_gmail.models.google_gmail_mixin.GoogleGmailMixin._fetch_gmail_access_token",
+            return_value=("new-access-token", int(datetime(2021, 12, 15, 12, 0, 1).timestamp()))
+        ):
+            oauth2_string = self.mail_server._generate_oauth2_string("user-account", "refresh-token")
+        self.assertIn("new-access-token", oauth2_string)
```
Comment (reth-odoo) [inline addons/google_gmail/tests/test_google_gmail.py:49]: Usually we use `subTest`s if the test is short enough and similar enough. Avoids the overhead of set-up/clean-up.
Here as we don't modify the record it should be easy.

```py
original_access_token = self.mail_server.google_gmail_access_token
token_expiration = self.mail_server.google_gmail_access_token_expiration
date_times = [
    token_expiration - timedelta(seconds=10),
    token_expiration - timedelta(seconds=5),
    token_expiration + timedelta(seconds=1),
]
for current_datetime in date_times:
    new_access_token = f"new_token_{current_datetime.timestamp()}"
    new_access_token_expiration = int((current_datetime + timedelta(hours=1)).timestamp())
    with self.subTest(currenct_datetime=current_datetime), mock.patch(
"odoo.addons.google_gmail.models.google_gmail_mixin.GoogleGmailMixin._fetch_gmail_access_token",
 return_value=(new_access_token, new_access_token_expiration)
        ):
            oauth2_string = self.mail_server._generate_oauth2_string("user-account", "refresh-token")
            self.assertIn(new_access_token, oauth2_string)  # would use a strict equality here if the result is predictable/not much longer
```

**Review by reth-odoo — COMMENTED** [review]
@petrus-v I can do the alarm :alarm_clock:  ^^
We're done with our busy period, thanks for your patience.

Here's some comments, we can also cherry-pick your commit into a local branch if you'd prefer. (saving you the back and forth of review)

File: addons/google_gmail/tests/test_google_gmail.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+from unittest import mock
+from datetime import datetime
+from freezegun import freeze_time
+
+
+class TestIrMailServer(TransactionCase):
+
+    def setUp(self):
+        super().setUp()
+        self.mail_server = self.env["ir.mail_server"].new(
```
[the remark this answers is not in the document]
Comment (petrus-v) [reply addons/google_gmail/tests/test_google_gmail.py]: done !

File: addons/google_gmail/tests/test_google_gmail.py [hunk 9]
```diff
@@ -0,0 +1,49 @@
+from odoo.tests.common import TransactionCase
+from unittest import mock
+from datetime import datetime
+from freezegun import freeze_time
+
+
+class TestIrMailServer(TransactionCase):
+
+    def setUp(self):
```
[the remark this answers is not in the document]
Comment (petrus-v) [reply addons/google_gmail/tests/test_google_gmail.py]: done !

File: addons/google_gmail/tests/test_google_gmail.py (hunk 7 above)
Comment (petrus-v) [reply addons/google_gmail/tests/test_google_gmail.py]: I don't think this make the code easier to maintains, here my point of view:
* I'm not sure it makes tests more readable
* I did it because I'd like to see this PR to be merged at some point, so here a PR using subTest !
* Your suggestion do not test the same as the code I've written previously, because:
  *  you don't freeze the time so it would have always call the `_fetch_gmail_access_token` method  while my fist test don't.
  * the state of self.google_gmail_access_token_expiration change between each calls that make it no verry predictable
* `_generate_oauth2_string` change the state of the self, between 2 subTests the setUp doen't seems to be called so I had to setup to properly set `google_gmail_access_token_expiration` attribute inside the method before each calls.

I've add assertion on logging message to make it more explicit which part of the code is expected to be call.

I've done this change in a separate commit for the time being to helps with the review, let me know if I should squash it with the proper one if you are happy with it?

Comment (reth-odoo) [reply addons/google_gmail/tests/test_google_gmail.py]: @tde-banana-odoo want to take a look at this one again? ^ I'll let you choose whether to squash or drop the "use subtests" commit

Comment (reth-odoo) [reply addons/google_gmail/tests/test_google_gmail.py]: @tde-banana-odoo :owl:

**Comment by tde-banana-odoo** [comment]
@robodoo delegate=@reth-odoo 

Too much stuff to check so delegating :)

File: addons/google_gmail/tests/test_google_gmail.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+        cls.mail_server = cls.env["ir.mail_server"].create(
+            {
+                "name": "Gmail",
+                "smtp_host": "fake.host",
+                "google_gmail_access_token": "fake_access_token",
+            }
+        )
+
+    def test_generate_oauth2_string_token(self):
+
```
Comment (reth-odoo) [inline addons/google_gmail/tests/test_google_gmail.py:21]: Would be nice to have a docstring

File: addons/google_gmail/tests/test_google_gmail.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+                self.assertEqual(
+                    f"user=user-account\1auth=Bearer {expected_token}\1\1",
+                    oauth2_string,
+                )
+                if assert_new_token_generation_called:
+                    mock_fetch_gmail_access_token.assert_called_once()
+                else:
+                    mock_fetch_gmail_access_token.assert_not_called()
+
+                mock_logger.assert_called_once_with(*expected_log)
```
Comment (reth-odoo) [inline addons/google_gmail/tests/test_google_gmail.py:75]: fyi `assertLogs` exists, but this works

File: addons/google_gmail/tests/test_google_gmail.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+                current_datetime
+            ), mock.patch(
+                "odoo.addons.google_gmail.models.google_gmail_mixin._logger.info"
+            ) as mock_logger, mock.patch(
+                "odoo.addons.google_gmail.models.google_gmail_mixin.GoogleGmailMixin._fetch_gmail_access_token",
+                return_value=(
+                    "new-access-token",
+                    int(datetime(2021, 12, 15, 12, 0, 1).timestamp()),
+                ),
+            ) as mock_fetch_gmail_access_token:
```
Comment (reth-odoo) [inline addons/google_gmail/tests/test_google_gmail.py:59]: Using overarching parentheses and inlining what can be seems more readable to me
```suggestion
with (
    self.subTest(currenct_datetime=current_datetime),
    freeze_time(current_datetime),
    mock.patch("odoo.addons.google_gmail.models.google_gmail_mixin._logger.info") as mock_logger,
    mock.patch(
        "odoo.addons.google_gmail.models.google_gmail_mixin.GoogleGmailMixin._fetch_gmail_access_token",
        return_value=("new-access-token", int(datetime(2021, 12, 15, 12, 0, 1).timestamp()))
    ) as mock_fetch_gmail_access_token
):
```

File: addons/google_gmail/tests/test_google_gmail.py [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+                self.mail_server.google_gmail_access_token_expiration = int(
+                    datetime(2021, 12, 15, 11, 0, 0).timestamp()
+                )
```
Comment (reth-odoo) [inline addons/google_gmail/tests/test_google_gmail.py:62]: I would set the unchanging timestamps at the start of the test as `new_token_expiry` (for the patch) and `current_token_expiry` (here). To "set the scene".

**Review by reth-odoo — COMMENTED** [review]
Thanks
Alright then, let's try to get this merged next week at last :pray: 
@petrus-v 

Can you fixup the "subtest" commit into the original one?
And add a self-contained explanation in the commit message of why the change was done? It's best if we don't *have* to follow links
Finally, would also elaborate a bit on the last commit message. e.g. "Disconnection errors do not really matter once the mail sending flow is over. We now ignore those, which prevents disconnection errors from potentially overshadowing more useful errors that may have been thrown prior."

Also a few more comments while we're here

File: addons/google_gmail/tests/test_google_gmail.py (hunk 11 above)
Comment (petrus-v) [reply addons/google_gmail/tests/test_google_gmail.py:75]: thanks for tips, I'm wondering if assertLogs depends on logging configuration, I've tried but didn't get any INFO logs so I kept code in place

File: addons/google_gmail/tests/test_google_gmail.py (hunk 12 above)
Comment (petrus-v) [reply addons/google_gmail/tests/test_google_gmail.py]: done

**Review by reth-odoo — APPROVED** [review]
Tested locally, everything seems right to me. Thanks for the fix, and your patience :p

bibidibabid-ont-break-everything-please
@robodoo r+
